base/show.jl

```diff
@@ -1411,6 +1431,9 @@ end
 """
 function print_matrix_repr(io, X::AbstractArray)
     compact, prefix = array_eltype_show_how(X)
+    if compact && !haskey(io, :compact)
+        io = IOContext(io, :compact => compact)
```

Comment:
I think we might want to keep this function type-stable? I didn't benchmark the possible added performance penalty of making the printing routines non-statically-typed on the IO argument in the initial PR for IOContext.